### ISSUE TITLE
Leaked Static SQLiteDatabase in SqlTileWriter

### ISSUE THREAD

**Issue by jordantrain** [osmdroid/osmdroid#1536]
## Issue Type

[X] Question
[X] Bug
[ ] Improvement
[ ] Build system related
[ ] Performance
[ ] Documentation

## Description and/or steps/code to reproduce the problem:
In version 6.16.
Why is the SQLiteDatabase in SqlTileWriter a static variable?  This causes a huge memory leak that lasts until the application process is destroyed which may be long after the application is closed.  You can easily visualize the leak by deleting the "cache.db" file and all map tiles in your osmdroid cache file and removing any other tile sources that the app may be using.  Turn off internet and restart the app and the leaked tiles from the static SQLiteDatabase are visible.

**Comment by monsieurtanuki** [comment]
I haven't had a look at the SqlTileWriter code recently.

I guess the point of making it `static` is more or less to have a "permanent" database.

For instance, when you display a new map, some tiles are in the database and some tiles are being downloaded then stored _asynchronously_: we cannot say "oh the user's just switched to landscape mode therefore we close the database and reopen it with the new Activity, it's too bad that I wanted to save the tile into the database at the very same moment".
And regarding reads, I guess keeping the database open makes the very first access quicker. And we constantly read the tiles from the database - as we keep the memory cache as little as possible.

In short: we need to have the database constantly open, for reads and writes.
Should we dedicate:
* a distinct instance of SqlTileWriter to the asynchronous download processes (WRITE) ?
* an instance of SqlTileWriter to each map view (READ)?

What are the best practices around that? What are your suggestions?

**Comment by jordantrain** [comment]
Ideally, getting a handle on the SqlTileWriter and closing the database when the application closes would work.  Unfortunately, android does not notify the Application object when the app closes so that is not an option.  I wouldn't have a problem with:  "oh the user's just switched to landscape mode therefore we close the database and reopen it with the new Activity, it's too bad that I wanted to save the tile into the database at the very same moment".  If the writes are wrapped in a try-catch block, let the write fail and no harm done.  The way it is now, all tiles stored during the MapView's life remain in memory after the MapView's hosting Activity is destroyed.

**Comment by monsieurtanuki** [comment]
Hey, you edited your comment while I was answering ;)

> no harm done

Except that we've just downloaded a tile, for nothing, and that's precisely one tile we were currently asking for. Not nice for people with low bandwidth.

> The way it is now, all tiles stored during the MapView's life remain in memory after the MapView's hosting Activity is destroyed.

I haven't tried your OP leak detection test yet.

What about my suggestion above:
* a distinct instance of SqlTileWriter for the asynchronous download processes (WRITE), closed after x seconds/minutes of inactivity, and opened on demand
* an instance of SqlTileWriter for each map view, to be closed on detach (READ)

**Comment by jordantrain** [comment]
I think "an instance of SqlTileWriter for each map view, to be closed on detach (READ)", but would this be any different or better than closing the database in the SqlTileWriters' onDetach method?

**Comment by monsieurtanuki** [comment]
I've just read things on SO and could not find anything 100% clear about if or when the database should be closed.
If you find something elegant I'd be glad to have a look at it and/or to implement it.
Btw there's leakcanary in osmdroid/dev, and there are obviously leaks in the demo app but they don't seem to be related to the database.

**Comment by jordantrain** [comment]
Go ahead and try the steps to reproduce the problem I mentioned and see what you think.

**Comment by spyhunter99** [comment]
i really don't remember why it was static. It may have been because of potentially multiple instances of SqlTileWriter with a tile provider chain. IIRC you can't open the same database twice or perhaps it was a concurrency issue.  It's probably worth the time to clean this up.

**Comment by monsieurtanuki** [comment]
@spyhunter99 We need the database to be open all the time, even while you're rotating your smartphone, because we're downloading tiles at any time.
